### ISSUE TITLE
Bug: parsing error for optional access and increment operator

### ISSUE THREAD

**Issue by SimenB** [typescript-eslint/typescript-eslint#8232]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=MYewdgzgLgBAlpKBDMwCmMC8MwFcA2%2BA3AFAkLQroD8AdAEYi5gAmA1G0UA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
instance?.bound++;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
  }
}
```

### Expected Result

No parsing error

### Actual Result

`Parsing error: Invalid left-hand side expression in unary operation 3:1 - 3:2`

### Additional Info

Essentially same as #8201

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `N/A` |
| `@typescript-eslint/parser`            | `6.18.1` |
| `@typescript-eslint/rule-tester`       | `N/A` |
| `@typescript-eslint/scope-manager`     | `N/A` |
| `@typescript-eslint/typescript-estree` | `N/A` |
| `@typescript-eslint/type-utils`        | `N/A` |
| `@typescript-eslint/utils`             | `N/A` |
| `TypeScript`                           | `5.3.3` |
| `ESLint`                               | `8.48.0` |
| `node`                                 | `N/A` |

**Comment by auvred** [comment]
This is a semantic error in TS: [TS Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gS3RqKYIAXPAN4BGSArisKStQLYUgwC+A3AFDe6YEiAfgB0VWsADUkzkA)
> The operand of an increment or decrement operator may not be an optional property access.


---


Also `espree` reports it as error [ESLint Playground](https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgaW5zdGFuY2UgPSBudWxsO1xuXG5pbnN0YW5jZT8uYm91bmQrKzsiLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)

> Parsing error: Optional chaining cannot appear in left-hand side

**Comment by Josh-Cena** [comment]
I'm on vacay so can't check too deeply. Yes, using a non-null assertion should be valid and is fixed by #8201. However using optional chaining is definitely not supported by any other tooling. Are you using a Babel plugin to compile that code?

**Comment by SimenB** [comment]
We've had it since March 2021 in tests: https://github.com/jest-community/eslint-plugin-jest/commit/b1f4ed3f6bb0264fdefb5138ba913fa2bacc725c#diff-00dd5fa94234729ac1f6fe1595dad705585a9d51987bab4d47f16b48cb96c0f7R402

**Comment by Josh-Cena** [comment]
Ah, so that code is essentially an inlined mock input that happens to be invalid code which we just didn't handle before. I suggest you remove that example because no one will expect it to be linted successfully in the first place.

**Comment by Josh-Cena** [comment]
Closing because it is, in fact, a syntax error that isn't meant to be supported by any analyzer

**Comment by SimenB** [comment]
Fair 👍 Would be nice with a better error message (like all of TS, espree and runtime gives), but no biggie
